lib/awful/screenshot.lua

```diff
@@ -0,0 +1,870 @@
+
+      if date_start and date_end then
+        return directory, string.sub(base_name, 1, date_start - 1)
+      end
+
+      return directory
+
+    end
+
+  end
+
+  return nil
+
+end
+
+-- Internal function to configure the directory/prefix pair for any call to the
+-- screenshot API.
+--
+-- This supports using a different directory or filename prefix for a particular
+-- use by passing the optional arguments. In the case of an initalized object
+-- with  no arguments passed, the stored configuration parameters are quickly
+-- returned. This also technically allows the user to take a screenshot in an
+-- unitialized state and without passing any arguments.
+local function configure_path(directory, prefix)
+
+  local _directory, _prefix
+
+  if not initialized or directory then
+    _directory = check_directory(directory)
+    if not _directory then
+      return
+    end
+  else
+    _directory = module_default_directory
+  end
+
+
+  if not initialized or prefix then
+    _prefix = check_prefix(prefix)
+    if not _prefix then 
+      return
+    end
+  else
+    _prefix = module_default_prefix
+  end
+
+  -- In the case of taking a screenshot in an unitialized state, store the
+  -- configuration parameters and toggle to initialized going forward.
+  if not initialized then
+    module_default_directory = _directory
+    module_default_prefix = _prefix
+    initilialized = true
+  end
+  
+  return _directory, _prefix
+
+end
+
+local function make_filepath(directory, prefix)
+  local _directory, _prefix
+  local date_time = tostring(os.date("%Y%m%d%H%M%S"))
+  _directory, _prefix = configure_path(directory, prefix)
+  local filepath = _directory .. _prefix .. date_time .. ".png"
+  return _directory, _prefix, filepath
+end
+
+-- Internal function to do the actual work of taking a cropped screenshot
+--
+-- This function does not do any data checking. Public functions should do
+-- all data checking. This function was made to combine the interactive sipper
+-- run by the mousegrabber and the programmatically defined snip function,
+-- though there may be other uses.
+local function crop_shot(x, y, width, height)
+
+  local source, target, cr
+
+  source = gears.surface(root.content())
+  target = source:create_similar(cairo.Content.COLOR,
+                         width, height)
+
+  cr = cairo.Context(target)
+  cr:set_source_surface(source, -x, -y)
+  cr:rectangle(0, 0, width, height)
+  cr:fill()
+
+  return target
+
+end
+
+
+-- Internal function used by the snipper mousegrabber to update the frame outline
+-- of the current state of the cropped screenshot
+--
+-- This function is largely a copy of awful.mouse.snap.show_placeholder(geo),
+-- so it is probably a good idea to create a utility routine to handle both use
+-- cases. It did not seem appropriate to make that change as a part of the pull
+-- request that was creating the screenshot API, as it would clutter and
+-- confuse the change.
+function show_frame(ss, geo)
```

Comment:
Please add `local ` in front (here and in a few other places)

Reply:
@Elv13, just to keep the Lua scoping rules correct, I should do this for top level module functions but not functions that are elements of tables. So I would want:

```
local function show_frame(ss, geo)
```

but

```
function screenshow_methods.root(ss)
```

Is this right?

Reply:
```lua
local function screenshot_methods.root(ss) end
```
would throw a syntax error. Just like the equivalent would

```lua
local screenshot_methods = {
  local root = function(ss) end
}
```